backend/courses/management/commands/recompute_topics.py

```diff
@@ -0,0 +1,104 @@
+from django.core.management.base import BaseCommand
+from django.db import transaction
+from django.db.models import Q, Subquery
+from tqdm import tqdm
+
+from courses.models import Course, Topic
+from courses.util import all_semesters
+
+
+def garbage_collect_topics():
+    """
+    Deletes topics with no courses.
+    """
+    Topic.objects.filter(
+        ~Q(id__in=Subquery(Topic.objects.filter(courses__isnull=False).values("id"))),
+    ).delete()
+
+
+def recompute_topics(min_semester: str = None, verbose=False):
+    """
+    Course topics are directly derived from the `Course.parent_course` graph.
+        - Any course without a parent gets its own topic.
+        - Any single child of a parent inherits the parent's topic.
+        - Any child with siblings that exactly matches their parent in full_code
+          and title inherits the parent's topic.
+        - Any child with siblings that does not exactly match their parent
+          gets its own topic.
+    These rules are applied sequentially in increasing order of semester.
+
+    :param min_semester: The min semester from which to recompute topics. This function
+        will reset topics for courses from all semesters since min_semester.
+    :param verbose: Whether to print status/progress updates.
+    """
+    semesters = [sem for sem in all_semesters() if not min_semester or sem >= min_semester]
+    if verbose:
+        print("Recomputing topics from the parent_course graph.")
+    for i, semester in enumerate(sorted(semesters)):
+        if verbose:
+            print(f"Processing semester {semester} ({i+1}/{len(semesters)})...")
+        with transaction.atomic():
+            courses = list(
+                Course.objects.filter(semester=semester)
+                .select_related("parent_course")
+                .select_related("parent_course__topic")
+                .prefetch_related("parent_course__children")
+            )
+            courses_to_update = []  # to update .topic
+            topics_to_update = []  # to update .most_recent
+            for course in tqdm(courses):
+                parent = course.parent_course
+                if not parent:
+                    course.topic = Topic(most_recent=course)
+                    course.topic.save()
+                    continue
+                if (
+                    parent.full_code == course.full_code
```

Comment:
This is speculative, but do you know how the crosswalk handles crosslistings? Seems like if it includes mappings from the same 3 digit code to multiple crosslistings we could end up with the topic explosion we were hoping to avoid...

Reply:
Wow, GREAT catch. Just checked the crosswalk and it has the confusing behavior of mapping a single old code to all the new crosslistings, rather than mapping crosslistings separately like `OIDD-291 -> OIDD-2910` and `MGMT-291 -> MGMT-2910`. Modified the load crosswalk script to split up crosslistings appropriately.